README.md

```diff
@@ -0,0 +1,453 @@
+  and use strict equality to compare input
+*/
+const myQuery = queryJSON(fetchFunction, expire(10000));
+```
+
+Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as a `key` to store the result obtained:
+```
+// usersCache is empty
+usersCache: {}
+
+//a user is fetched
+getUser({ userId: 1 }) -> { userName: "Mario" }
+
+// usersCache is now populated
+usersCache: {
+  [{ userId: 1 }]: { userName: Mario }
+}
+```
+
+From that moment onwards, when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will:
+
+1. attempt to retrieve data from the [**`Cache`**](#Cache)
+2. match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).
+
+If the result is valid it is returned, otherwise the `Fetch` function will be re-run in order to get new and valid data. The two flows are relatively simple:
+##### Valid CacheValue
+!["cached flow"](docs/CachedValue.svg)
+
+##### Invalid CacheValue
+when you call `run` or `subscribe` on a `query` with a combination of `inputs` that was never used before (or whose last use ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
+!["cached flow"](docs/UncachedOrErrorValue.svg)
+
+There are two ways to get a query result:
+
+```ts
+type Error = "500" | "404";
+type User = { userName: String };
+
+declare function getUser(userId: number): TaskEither<Error, User>
+
+const userQuery: CachedQuery<number, Error, User> = query(
+  getUser,
+  refetch
+)(refetch);
+
+declare function dispatchError(e: Error): void;
+declare function setCurrentUser(e: User): void;
+
+// feeding your query to `observe` will give you an observable on the query
+// N.B. until now no fetch is yet attempted, avenger will wait until the first subscription is issued
+const observable: Observable<QueryResult<Error, User>> = observe(userQuery);
+
+// this will trigger the fetch function
+observable.subscribe(dispatchError, setCurrentUser);
+
+// alternatively you can call `run` on your query and it will return a TaskEither<Error, User>
+// you can then use it imperatively
+const task: TaskEither<Error, User> = userQuery.run(1);
+const result: Either<Error, User> = await task.run();
+```
+
+although the `run` method is available to check a query result imperatively, it is highly suggested the use of the `observe` utility in order to be notified in real time of when data changes.
+
+Either way, whenever you ask for a query result you will end up with an object with the [**`QueryResult`**](#QueryResult) signature that conveniently lets you fold to decide the best way to handle the result. The `fold` method takes 3 functions as parameters, the first is used to handle a `Loading` result, the second is used in case a `Failure` occurs and the last one handles `Success` values.
```

Comment:
takes ~~3~~ **three** functions as parameters **colon** the first ... **semicolon** the second ...